File: examples/java-junit-maven/src/test/java/examples/TablesAndDocStringsSteps.java
package examples;

import dev.varar.Registrar;
import dev.varar.State;
import dev.varar.StateBinder;
import dev.varar.StepDefinitions;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public final class TablesAndDocStringsSteps implements StepDefinitions {

    record Ctx() implements State {}

    @Override
    public void defineSteps(Registrar registrar) {
        StateBinder<Ctx> s = registrar.steps(Ctx::new);

        // Whole-table mode: the table arrives as List<List<String>> (header row
        // first). It is this sensor's only slot, so return the reproduced table
        // bare — Vár compares every cell.
        s.sensor("Uppercase each one:", (Ctx ctx, List<List<String>> rows) -> {
            List<Map<String, String>> out = new ArrayList<>();
            for (List<String> row : rows.subList(1, rows.size())) {
                out.add(Map.of("before", row.get(0), "after", row.get(0).toUpperCase(Locale.ROOT)));
            }
            return out;
        });

        // Doc-string mode: two slots ({word} plus the trailing doc string), so
        // return one element per slot.
        s.sensor("Greet {word}:", (Ctx ctx, String name, String doc) -> List.of(name, "Hello, " + name + "!\n"));
    }
}

File: examples/kotlin-junit/src/test/kotlin/examples/tables-and-docstrings.steps.kt
@file:JvmName("TablesAndDocStringsSteps")

package examples

import dev.varar.kotlin.sensor
import dev.varar.kotlin.steps

val tablesAndDocStringsSteps = steps {
    // Whole-table mode: the table arrives as List<List<String>> (header row
    // first). It is this sensor's only slot, so return the reproduced table
    // bare — Vár compares every cell.
    sensor("Uppercase each one:") { rows: List<List<String>> ->
        rows.drop(1).map { row -> mapOf("before" to row[0], "after" to row[0].uppercase()) }
    }
    // Doc-string mode: two slots ({word} plus the trailing doc string), so
    // return one element per slot.
    sensor("Greet {word}:") { name: String, _: String ->
        listOf(name, "Hello, $name!\n")
    }
}
